File: lib/src/state/contract.ts
import * as borsh from "borsh";
import { 
    Buffer 
} from "buffer";
import { 
    PublicKey 
} from "@solana/web3.js";


export class Contract {

    authority: PublicKey;
    challenge: PublicKey;
    earner: PublicKey;
    event: PublicKey;
    escrows_created: number;
    status: number;
    bump: number;

    constructor(props: {
        authority: PublicKey,
        challenge: PublicKey,
        earner: PublicKey,
        event: PublicKey,
        escrows_created: number,
        status: number,
        bump: number,
    }) {
        this.authority = props.authority;
        this.challenge = props.challenge;
        this.earner = props.earner;
        this.event = props.event;
        this.escrows_created = props.escrows_created;
        this.status = props.status;
        this.bump = props.bump;
    }

    toBuffer() { 
        return Buffer.from(borsh.serialize(ContractSchema, this)) 
    }
    
    static fromBuffer(buffer: Buffer) {
        const contract = borsh.deserialize(ContractSchema, Contract, buffer);
        return new Contract({
            authority: contract.authority,
            challenge: contract.challenge,
            earner: contract.earner,
            event: contract.event,
            escrows_created: contract.escrows_created,
            status: contract.status,
            bump: contract.bump,
        });
    }
}

export const ContractSchema = new Map([
    [ Contract, { 
        kind: 'struct', 
        fields: [ 
            ['authority', [32]],
            ['challenge', [32]],
            ['earner', [32]],
            ['event', [32]],
            ['escrows_created', 'u32'],
            ['status', 'u8'],
            ['bump', 'u8'],
        ],
    }]
]);

File: program/src/state/contract.rs
use borsh::{ BorshDeserialize, BorshSerialize };
use solana_program::pubkey::Pubkey;

#[derive(BorshDeserialize, BorshSerialize)]
pub struct Contract {
    pub authority: Pubkey,
    pub challenge: Pubkey,
    pub earner: Pubkey,
    pub event: Pubkey,
    pub expiration: String,
    pub escrows_created: u32,
    pub status: u8, // 0: closed, 1: open, 2: fulfilled
    pub bump: u8,
}


impl Contract {

    pub const SEED_PREFIX: &'static str = "contract";
    
    pub fn new(
        authority: Pubkey,
        challenge: Pubkey,
        earner: Pubkey,
        event: Pubkey,
        expiration: String,
        bump: u8,
    ) -> Self {

        Contract {
            authority,
            challenge,
            earner,
            event,
            expiration,
            escrows_created: 0,
            status: 1,
            bump,
        }
    }
}